engine/crates/runtime-local/src/rate_limiting/in_memory/key_based.rs

```diff
@@ -34,48 +37,73 @@ impl RateLimiterContext for RateLimitingContext {
     }
 }
 
-#[derive(Default)]
 pub struct InMemoryRateLimiter {
-    inner: HashMap<String, governor::DefaultKeyedRateLimiter<usize>>,
+    limiters: Arc<RwLock<HashMap<String, governor::DefaultKeyedRateLimiter<usize>>>>,
 }
 
 impl InMemoryRateLimiter {
-    pub fn runtime(config: KeyedRateLimitConfig<'_>) -> RateLimiter {
-        let mut limiter = Self::default();
+    pub fn runtime(
+        config: KeyedRateLimitConfig,
+        mut updates: mpsc::Receiver<HashMap<String, GraphRateLimit>>,
```

Comment:
Have you considered using a Watch channel like Redis implementation? You would just need to use https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed AFAIK to only handle the last event. Intermediate events have no value after all